lib/semantics/resolve-names.cc

```diff
@@ -2848,39 +2856,45 @@ void DeclarationVisitor::Post(const parser::DerivedTypeDef &x) {
       Say(stmt.source,
           "A sequence type may not have type parameters"_err_en_US);  // C740
     }
-    if (derivedTypeInfo_.sawContains) {
-      Say(stmt.source,
-          "A sequence type may not have a CONTAINS statement"_err_en_US);  // C740
-    }
   }
+  Walk(std::get<std::list<parser::Statement<parser::ComponentDefStmt>>>(x.t));
+  Walk(std::get<std::optional<parser::TypeBoundProcedurePart>>(x.t));
+  Walk(std::get<parser::Statement<parser::EndTypeStmt>>(x.t));
   derivedTypeInfo_ = {};
   PopScope();
+  return false;
 }
 bool DeclarationVisitor::Pre(const parser::DerivedTypeStmt &x) {
   return BeginAttrs();
 }
 void DeclarationVisitor::Post(const parser::DerivedTypeStmt &x) {
+  // Resolve the EXTENDS() clause before creating the derived
+  // type's symbol to foil attempts to recursively extend a type.
+  auto *extendsName{derivedTypeInfo_.extends};
+  const Symbol *extendsType{nullptr};
+  if (extendsName != nullptr) {
```

Comment:
Can we check here for extendsName being the same as the derived type name and provide a better error, like "derived type cannot extend itself"?

Reply:
Sure.